flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TemporalRowtimeJoin.scala

```diff
@@ -0,0 +1,339 @@
+    if (lastUnprocessedTime < Long.MaxValue) {
+      registerTimer(lastUnprocessedTime)
+    }
+  }
+
+  /**
+    * @return a row time of the oldest unprocessed probe record or Long.MaxValue, if all records
+    *         have been processed.
+    */
+  private def emitResultAndCleanUpState(timerTimestamp: Long): Long = {
+    val rightRowsSorted = getRightRowsSorted(rightRowtimeComparator)
+    var lastUnprocessedTime = Long.MaxValue
+
+    val leftIterator = leftState.entries().iterator()
+    while (leftIterator.hasNext) {
+      val leftEntry = leftIterator.next()
+      val leftRow = leftEntry.getValue
+      val leftTime = getLeftTime(leftRow)
+
+      if (leftTime <= timerTimestamp) {
+        val rightRowIndex = latestRightRowToJoin(rightRowsSorted, leftTime)
+
+        if (rightRowIndex >= 0) {
+          val rightRow = rightRowsSorted.get(rightRowIndex)
+
+          joinFunction.join(leftRow, rightRow, cRowWrapper)
+        }
+        leftIterator.remove()
+      }
+      else {
+        lastUnprocessedTime = Math.min(lastUnprocessedTime, leftTime)
+      }
+    }
+
+    cleanUpState(timerTimestamp, rightRowsSorted)
+    lastUnprocessedTime
+  }
+
+  /**
+    * Removes all right entries older then the watermark, except the latest one. For example with:
+    * rightState = [1, 5, 9]
+    * and
+    * watermark = 6
+    * we can not remove "5" from rightState, because left elements with rowtime of 7 or 8 could
+    * be joined with it later
+    */
+  private def cleanUpState(timerTimestamp: Long, rightRowsSorted: util.List[Row]) = {
+    var i = 0
+    while (i < firstIndexToKeep(timerTimestamp, rightRowsSorted)) {
+      val rightTime = getRightTime(rightRowsSorted.get(i))
+      rightState.remove(rightTime)
+      i += 1
+    }
+  }
+
+  private def firstIndexToKeep(timerTimestamp: Long, rightRowsSorted: util.List[Row]): Int = {
+    val firstIndexNewerThenTimer =
+      indexOf[Row](rightRowsSorted, row => getRightTime(row) > timerTimestamp)
+
+    if (firstIndexNewerThenTimer < 0) {
+      rightRowsSorted.size() - 1
+    }
+    else {
+      firstIndexNewerThenTimer - 1
+    }
+  }
+
+  private def indexOf[T](list: util.List[T], predicate: T => Boolean): Int = {
+    val iter = list.listIterator
+    while (iter.hasNext) {
+      if (predicate(iter.next)) {
+        return iter.previousIndex
+      }
+    }
+    -1
+  }
+
+  /**
+    * Binary search `rightRowsSorted` to find the latest right row to join with `leftTime`.
+    * Latest means a right row with largest time that is still smaller or equal to `leftTime`.
+    *
+    * @return index of such element. If such row was not found (either `rightRowsSorted` is empty
+    *         or all `rightRowsSorted` are are newer) return -1.
+    */
+  private def latestRightRowToJoin(rightRowsSorted: util.List[Row], leftTime: Long): Int = {
+    latestRightRowToJoin(rightRowsSorted, 0, rightRowsSorted.size - 1, leftTime)
+  }
+
+  private def latestRightRowToJoin(
```

Comment:
This method it seems that it could return the element itself and not its index. This would save one state access.

Reply:
Yes, good point I will change it.